File: Desktop/dsa/Search and Sort/quickSort.cpp
#include<iostream>
using namespace std;

void quickSort(int* A, const int n){

	if(n <= 1) return;
	int pivot = A[0], less_than_p = 0;

	for(int scanner = 1; scanner < n; scanner++){
		if(A[scanner] < pivot){
			int temp = A[scanner];
			A[scanner] = A[less_than_p+1];
			A[less_than_p+1] = temp;

			less_than_p++;
		}
	}


	A[0] = A[less_than_p];
	A[less_than_p] = pivot;

	quickSort(A, less_than_p);
	quickSort(A+less_than_p+1, n-less_than_p-1);
	
}

int main(){
	int A[] = {0, 4, 3, 2, 5, 6, 3, 5};
	
	int n = sizeof(A)/4;
	
	quickSort(A,n);
	for(int i = 0; i < n; i++){
		cout<<A[i]<<" ";
	}
	cout<<endl;

	return 0;
}

File: Desktop/dsa/Search and Sort/quickSort.py
def quickSort(A):
	n = len(A)
	pivot = A[0]

	less_than_p = 0
	
	for i in range(n-1):
		
		scan_index = i+1
		if A[scan_index] < pivot:
			temp = A[less_than_p+1]
			A[less_than_p+1] = A[scan_index]
			A[scan_index] = temp
			less_than_p += 1
	
	A[0] = A[less_than_p]
	A[less_than_p] = pivot
	if less_than_p > 0:
		A[0:less_than_p] = quickSort(A[0:less_than_p])
	if less_than_p+1 < n:
		A[less_than_p+1:n] = quickSort(A[less_than_p+1:n])
	
	return A
	
	
A = [3, 2, 6, 2, 6, 5, 3, 8]
B = quickSort(A)
print B